### ISSUE TITLE
NIL is not a NUMBER in CL-SAT:PRINT-CNF

### ISSUE THREAD

**Issue by phmarek** [guicho271828/cl-sat#1]
Hi,

using some input with 1368 variables trying to `SAT:SOLVE` gives me (with both `minisat` and `glucose`) the error

```
The value
  NIL
is not of type
  NUMBER
   [Condition of type TYPE-ERROR]

Restarts:
  R 0.  RETRY - Retry SLIME REPL evaluation request.
  R 1. *ABORT - Return to SLIME's top level.
  R 2.  ABORT - abort thread (#<THREAD "repl-thread" RUNNING {1003061D83}>)

Frames:
  F 0.  (CL-SAT:PRINT-CNF #<CL-SAT:SAT-INSTANCE {1008EB7123}> #<SB-SYS:FD-STREAM for "file /tmp/cnf.eZcOsGA" {1009CE9013}>)
	File: ...quicklisp/software/cl-sat-20171130-git/src/3-print.lisp line 709
	Snippet:
	  (iter (for c in clauses)
	             (ematch c
	               ((list* 'or terms)
	                (format stream "~&c ~a" c)
	                (format stream "~&~{~a ~}0"
	                        (iter (for term in terms)
	                              (collect
	Locals:
	  ATOM:           ((OR (AND P0 ...
	  C:              (OR ..)
	  CLAUSES:        ((OR (...
	  #:CONS19:       (OR ..)
	  #:CONS3:        (AND (OR...
	  #:END-POINTER6: NIL
	  #:G181:         NIL
	  INSTANCE:       #<CL-SAT:SAT-INSTANCE {1008EB7123}>
	  #:IT2:          #<CL-SAT:SAT-INSTANCE {1008EB7123}>
	  #:IT20:         OR
	  #:IT4:          AND
	  #:LIST3:        NIL
	  #:LIST5:        NIL
	  #:RESULT4:      NIL
	  STREAM:         #<SB-SYS:FD-STREAM for "file /tmp/cnf.eZcOsGA" {1009CE9013}>
	  #:TEMP7:        NIL
	  TERM:           ((OR (AND...
	  TERMS:          (((OR (AND ...
	  VARIABLES:      (F170 T170 C170 D170 I170 A170 ...)
	  #:WHAT124:      (OR ..)
	  #:WHAT141:      ((OR (AND...
	  #:WHAT18:       #<CL-SAT:SAT-INSTANCE {1008EB7123}>

  F 1.  ((:METHOD CL-SAT:SOLVE (CL-SAT:SAT-INSTANCE T)) #<CL-SAT:SAT-INSTANCE {1008EB7123}> :MINISAT) [fast-method]
  F 2.  ((:METHOD CL-SAT:SOLVE (LIST T)) ((OR (AND P0 #1=# #1# #1# #1# ...) (AND #2=# L0 #2# #2# #2# ...) (AND #3=# #3# A0 #3# #3# ...) (AND #4=# #4# #4# I0 #4# ...) (AND #5=# #5# #5# #5# D0 ...) ...) (OR (AN..
```

I'm using QL which currently provides `cl-sat-20171130-git`.

For a trivial example it works; but the expression I'm pushing in looks okay to me. What can I do to diagnose that problem?

**Comment by guicho271828** [comment]
https://github.com/guicho271828/cl-sat/blob/master/src/3-print.lisp#L31
I guess `position` is failing to find the element. Could you extract the full input?

**Comment by phmarek** [comment]
Did you receive my mail with the reproducer?
No need to hurry, I'd just like to know.

**Comment by guicho271828** [comment]
oops, sorry! I was not paying attention to the inbox recently

**Comment by guicho271828** [comment]
It was in the spam box, sorry! I will work on the weekend

**Comment by phmarek** [comment]
Thanks for looking - and your efforts!

**Comment by guicho271828** [comment]
Closing, as Tseyting transform is now added.

BTW, Tseytin transform produces a redundant SAT formula, and I observed that compacting the resulting formula actually *slows down* the computation. This is because the redundancy actually helps the solvers infer(learn) more clauses(constraints) for pruning. While the time used for checking the list of clauses/constraints inside the solver may take longer, the power of pruning outweighs the benefit of the shorter checklist.
